torch/distributed/checkpoint/examples/stateful_example.py

```diff
@@ -0,0 +1,104 @@
+# Owner(s): ["oncall: distributed"]
+
+
+import os
+import shutil
+
+import torch
+import torch.distributed as dist
+import torch.distributed.checkpoint as DCP
+import torch.multiprocessing as mp
+import torch.nn as nn
+from torch.distributed._tensor.device_mesh import init_device_mesh
+from torch.distributed.checkpoint.state_dict import (
+    _patch_model_state_dict,
+    _patch_optimizer_state_dict,
+)
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+
+
+CHECKPOINT_DIR = f"~/{os.environ['LOGNAME']}/checkpoint"
+
+
+class Model(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        torch.manual_seed(0)
+        self.net1 = nn.Sequential(nn.Linear(8, 16), nn.ReLU())
+        self.net2 = nn.Sequential(nn.Linear(16, 32), nn.ReLU())
+        self.net3 = nn.Linear(32, 64)
+        self.net4 = nn.Sequential(nn.ReLU(), nn.Linear(64, 8))
+
+    def forward(self, x):
+        return self.net4(self.net3(self.net2(self.net1(x))))
+
+    def get_input(self):
+        return torch.rand(8, 8, device="cuda")
+
+
+def _make_stateful(model, optim):
```

Comment:
what does this do?  should users rely on it, or should they write their own state function on their model?

Reply:
This question is a bit loaded, I'm still thinking about the best UX around this item in particular. In general, if users are defining objects with custom `Stateful` behavior, they should define `state_dict` and `load_state_dict` on those objects.

For model/optim, which need to call `get_state_dict` and `set_state_dict` in `state_dict` and `load_state_dict`, the two options so far are:

1. Create a wrapper:
    ```
    class DOptim:
           def __init__(self, model, optim):
               self.model = model
               self.optim = optim
            ...
    ```
We're still evaluating whether it makes sense to include the wrapper as part of DCP since it can be a little tricky and could lead to a less then ideal UX

2. The other option, which I don't think is a bad one, is using the `_patch` methods as above. The patch methods are still in testing but I think it's pretty reasonable